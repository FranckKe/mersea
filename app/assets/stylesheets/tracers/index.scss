.off-canvas-wrapper, .off-canvas-wrapper-inner {
height: 100%;
}

.report-link {
    color: red;
    padding-top: 1rem;
}

.traceur-title {
    text-align: center;
    padding-top: 1rem;
}

.tracers-flex-container {
    padding-bottom: 4rem;
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    flex-wrap: wrap;
    background-color: inherit;
    height: 100%;

    .card {
        flex-basis: 150px;
        flex-shrink: 1;
        flex-grow: 1;
        margin: .5rem;
        overflow: hidden;
        box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);

        .title {
            line-height: 3rem;
            font-size: 1.5rem;
            font-weight: 300;
        }
        .content {
            padding: .7rem;
            font-weight: 300;
            border-radius: 0 0 2px 2px;
        }
        p {
            margin: 0;
        }
        .action {
            border-top: 1px solid rgba(160, 160, 160, 0.2);
            padding: .7rem;
        }
        a {
            display: block;
            margin-right: 1.3rem;
            transition: color 0.3s ease;
            text-transform: uppercase;
            text-decoration: none;
            font-size: .8rem;
            word-wrap: break-word;
        }
        .image {
            position: relative;
            width: 100%;
            .title {
                position: absolute;
                bottom: 0;
                left: 0;
                padding: 1.3rem;
                color: #fff;
            }
            img {
                border-radius: 2px 2px 0 0;
            }
        }
    }
}