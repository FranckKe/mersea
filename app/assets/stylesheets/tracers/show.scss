.traceur-title {
  text-align: center;
  padding-top: 1rem;
}

.tracer-container {
  background-color: inherit;
  max-width: 1000px;
  margin-left: auto;
  margin-right: auto;
  padding-bottom: 4rem;

  .card {
    margin: 0.5rem;
    overflow: hidden;

    .title {
      line-height: 3rem;
      font-size: 1.5rem;
      font-weight: 300;
    }

    .description {
      padding: 0.7rem;
      font-weight: 300;
      border-radius: 0 0 2px 2px;
    }

    p {
      margin: 0;
    }

    .report-form {
      border-top: 1px solid rgba(160, 160, 160, 0.2);
      padding: 0.7rem;
    }

    a {
      color: red;
      transition: color 0.3s ease;
      text-transform: uppercase;
      text-decoration: none;
      font-size: 0.8rem;
      word-wrap: break-word;
    }

    .image {
      position: relative;

      .title {
        position: absolute;
        bottom: 0;
        left: 0;
        padding: 1.3rem;
        color: #fff;
      }

      img {
        border-radius: 2px 2px 0 0;
      }
    }
  }
}
