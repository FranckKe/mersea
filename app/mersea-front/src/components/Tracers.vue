<template>
  <section class="section">
    <h1>Tracers</h1>
    <div class="columns is-multiline is-mobile">
      <div class="column is-half-mobile is-one-third-tablet is-one-quarter-desktop" :key="idx" v-for="(tracer, idx) in tracers">
        <article class="card">
          <div class="card-image">
            <div class="img-wrapper" :style="'background-image: url(' + tracer.img + ')'">
            </div>
          </div>
          <div class="card-content">
            {{ tracer.title }}
          </div>
        </article>
      </div>
    </div>
  </section>
</template>


<script>
export default {
  name: 'tracers',
  data() {
    return {
      tracers: [
        {
          title: 'Hello1',
          content: 'Lorem ipsum lol',
          img: 'https://placeimg.com/1280/960/any'
        },
        {
          title: 'Hello2',
          content: 'Lorem ipsum lol',
          img: 'https://placeimg.com/1280/960/any'
        },
        {
          title: 'Hello3',
          content: 'Lorem ipsum lol',
          img: 'https://placeimg.com/1280/960/any'
        },
        {
          title: 'Hello4',
          content: 'Lorem ipsum lol',
          img: 'https://placeimg.com/1280/960/any'
        },
        {
          title: 'Hello5',
          content: 'Lorem ipsum lol',
          img: 'https://placeimg.com/1280/960/any'
        },
        {
          title: 'Hello6',
          content: 'Lorem ipsum lol',
          img: 'https://placeimg.com/1280/960/any'
        },
        {
          title: 'Hello7',
          content: 'Lorem ipsum lol',
          img: 'https://placeimg.com/1280/960/any'
        }
      ]
    }
  }
}
</script>
<style>
.card-image {
  height: 200px;
}

.img-wrapper {
  width: 100%;
  height: 100%;
  position: relative;
  float: left;
  background-position: 50% 50%;
  background-repeat: no-repeat;
  background-size: cover;
}
</style>
