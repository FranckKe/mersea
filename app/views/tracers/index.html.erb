<h1 class="traceur-title">Traceurs</h1>
<section class="tracers-flex-container">
    <% @tracers.each do |tracer| %>
        <div class="card">
            <%= link_to(tracer, { class: "image" }) do %>
                <%= image_tag tracer.photo.url %>
            <% end %>
            <div class="action">
                <%= link_to tracer.name, tracer %>
                <%= link_to "Reporter", tracer, class: "report-link" %>
            </div>
        </div>
        <% end %>
</section>