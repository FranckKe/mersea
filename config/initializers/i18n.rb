Rails.application.config.i18n.available_locales = %i[fr en es]
Rails.application.config.i18n.default_locale = 'fr'
