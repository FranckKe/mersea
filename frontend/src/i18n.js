import Vue from 'vue'
import VueI18n from 'vue-i18n'

Vue.use(VueI18n)

function loadLocaleMessages() {
  const locales = require.context(
    './locales',
    true,
    /[A-Za-z0-9-_,\s]+\.json$/i
  )
  const messages = {}
  locales.keys().forEach(key => {
    const matched = key.match(/([a-z0-9]+)\./i)
    if (matched && matched.length > 1) {
      const locale = matched[1]
      messages[locale] = locales(key)
    }
  })
  return messages
}

// Detect browser language and strip the "region" language
// en-GB => en
let browserI18n = navigator.language
if (navigator.language.indexOf('-') >= 0)
  browserI18n = navigator.language.substring(0, navigator.language.indexOf('-'))

// Default to english if browser language is not supported
let initializedLocal = ['en', 'es', 'fr'].includes(browserI18n)
  ? browserI18n
  : 'en'

export default new VueI18n({
  locale: initializedLocal,
  fallbackLocale: 'en',
  messages: loadLocaleMessages()
})
