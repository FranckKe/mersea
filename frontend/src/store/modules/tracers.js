import axios from 'axios'

const api = axios.create({
  baseURL: process.env.VUE_APP_API_URL,
  headers: { Accept: 'application/json', 'Content-Type': 'application/json' }
})

const state = {
  tracers: [],
  filteredTracers: [],
  loading: true,
  success: false,
  errors: [],
  displayFormat: 'grid',
  perPage: '10'
}

const getters = {
  getFilteredTracers: state => {
    return state.filteredTracers
  },
  getLoading: state => {
    return state.loading
  },
  getError: state => {
    return state.errors
  },
  getSuccess: state => {
    return state.success
  },
  getTracers: state => {
    return state.tracers
  },
  getPerPage: state => {
    return state.perPage
  },
  getTracerById: state => tracerId =>
    state.tracers.filter(tracer => tracerId === tracer.id)[0]
}

const mutations = {
  setTracers: (state, { tracers }) => {
    state.tracers = tracers
  },
  setLoading: (state, loading) => {
    state.loading = loading
  },
  setSuccess: state => {
    state.success = true
    state.loading = false
    state.errors = []
  },
  setError: (state, { errors }) => {
    state.errors = errors
    state.loading = false
  },
  setDisplayFormat: (state, format) => {
    state.displayFormat = format
  },
  setPerPage: (state, perPage) => {
    state.perPage = perPage
  },
  setFilteredTracers: (state, filteredTracers) => {
    state.filteredTracers = filteredTracers
  }
}

const actions = {
  async loadTracers({ commit }) {
    try {
      commit('setLoading', true)

      const tracersRes = await api.get(`/tracers`)
      const tracers = tracersRes.data
      commit('setTracers', { tracers })
      commit('setFilteredTracers', tracers.map(t => t.id))
      commit('setSuccess')
    } catch (e) {
      let errors = [e.message]
      commit('setError', { errors })
    }
  }
}

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations
}
