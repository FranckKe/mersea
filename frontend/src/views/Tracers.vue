<template>
  <div class="section">
    <div class="container">
      <h1 class="title is-1">{{ $tc('tracers', 2) | capitalize }}</h1>
      <div class="field">
        <div class="control has-icons-left">
          <span class="icon is-small is-left">
            <font-awesome-icon icon="search" />
          </span>
          <input class="input" type="text" v-bind:placeholder="$t('search')" v-model="searchKeywords">
        </div>
      </div>
      <div class="columns is-multiline is-mobile">
        <div class="column">
          <p class="has-text-left is-italic">{{ getFilteredTracers().length }} {{ $tc('tracers', getFilteredTracers().length) }} {{ $tc('displayed', getFilteredTracers().length) }} ({{ this.tracers.length }} {{ $t('total') }})</p>
        </div>
        <div class="column buttons is-one-half-mobile is-one-quarter-tablet is-one-quarter-desktop has-text-right">
          <button class="button" v-bind:class="[this.displayFormat === 'grid' ? 'is-primary' : '']" v-on:click="setDisplayFormat('grid')">
            <span class="icon is-small is-left">
              <font-awesome-icon icon="th-large" />
            </span>
          </button>
          <button class="button" v-bind:class="[this.displayFormat === 'list' ? 'is-primary' : '']" v-on:click="setDisplayFormat('list')">
            <span class="icon is-small is-left">
              <font-awesome-icon icon="th-list" />
            </span>
          </button>
        </div>
      </div>
      <transition name="fade" mode="out-in">
        <tracers-grid v-if="this.displayFormat === 'grid'" :tracers="getFilteredTracers()"></tracers-grid>
        <tracers-list v-if="this.displayFormat === 'list'" :tracers="getFilteredTracers()"></tracers-list>
      </transition>
    </div>
  </div>
</template>

<script>
import { createNamespacedHelpers } from 'vuex'
const {
  mapState,
  mapMutations,
  mapGetters,
  mapActions
} = createNamespacedHelpers('tracers')
import TracersGrid from '@/components/TracersGrid'
import TracersList from '@/components/TracersList'

export default {
  name: 'tracers',
  components: {
    TracersGrid,
    TracersList
  },
  data() {
    return {
      tracers: [],
      searchKeywords: ''
    }
  },
  async created() {
    try {
      await this.loadTracers()
    } catch (e) {
      this.$toast.open({
        message: this.$t('load_tracers_failure'),
        duration: 5000,
        type: 'is-danger'
      })
    }
    this.tracers = this.getData()
  },
  methods: {
    getFilteredTracers() {
      if (this.searchKeywords.trim() !== '') {
        let keywords = this.searchKeywords.toLowerCase()
        return this.tracers.filter(
          tracer => tracer.name.toLowerCase().indexOf(keywords) !== -1
        )
      } else {
        return this.tracers
      }
    },
    ...mapGetters(['getData']),
    ...mapMutations(['setDisplayFormat']),
    ...mapActions(['loadTracers'])
  },
  computed: {
    ...mapState({
      displayFormat: state => state.displayFormat,
      perPage: state => state.perPage
    })
  }
}
</script>
<style scoped>
div.section {
  padding-top: 0;
}
.card-image {
  height: 200px;
}

.heading_text {
  margin-bottom: 1.5rem;
}

.img-wrapper {
  width: 100%;
  height: 100%;
  position: relative;
  float: left;
  background-position: 50% 50%;
  background-repeat: no-repeat;
  background-size: cover;
}
</style>

<i18n>
{
  "en": {
    "tracers": "tracer | tracers",
    "search": "Search",
    "displayed": "displayed | displayed",
    "total": "total",
    "load_tracers_failure": "Failed to load tracers"
  },
  "fr": {
    "tracers": "tracer | tracers",
    "search": "Rechercher",
    "displayed": "affiché | affichés",
    "total": "en tout",
    "load_tracers_failure": "Échec de chargement des tracers"
  },
  "es": {
    "tracers": "trazadore | trazadores",
    "search": "Buscar",
    "displayed": "desplegado | desplegado",
    "total": "total",
    "load_tracers_failure": "Error al cargar los trazadores"
  }
}
</i18n>
