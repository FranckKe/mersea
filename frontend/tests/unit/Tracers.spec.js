import VueTestUtils from '@vue/test-utils'
import Tracers from '@/views/Tracers.vue'
import Vuex from 'vuex'
import tracers from '../../src/store/modules/tracers'
import Buefy from 'buefy'
import i18n from '../../src/i18n'
import { library } from '@fortawesome/fontawesome-svg-core'
import {
  faSearch,
  faThLarge,
  faThList
} from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faSearch, faThList, faThLarge)

VueTestUtils.config.mocks.$t = key => key
VueTestUtils.config.mocks.$tc = (key, count) => (key, count)

const localVue = VueTestUtils.createLocalVue()
localVue.use(Buefy)
localVue.use(Vuex)

localVue.component('font-awesome-icon', FontAwesomeIcon)
localVue.filter('capitalize', function(value) {
  if (!value) return ''
  value = value.toString()
  return value.charAt(0).toUpperCase() + value.slice(1)
})

describe('Tracers.vue', () => {
  let store

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        tracers: tracers
      }
    })
  })

  it('matches snapshot', () => {
    const wrapper = VueTestUtils.shallowMount(Tracers, {
      localVue,
      store,
      i18n
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
