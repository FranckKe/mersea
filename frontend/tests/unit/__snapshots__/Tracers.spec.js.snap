// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tracers.vue matches snapshot 1`] = `
<div class="section">
  <h1 class="heading_text is-size-3">Tracers</h1>
  <div class="field">
    <div class="control has-icons-left"><span class="icon is-small is-left"><font-awesome-icon-stub icon="search"></font-awesome-icon-stub></span>
      <input type="text" placeholder="Search for a tracer" class="input">
    </div>
  </div>
  <div class="columns is-multiline is-mobile">
    <div class="column">
      <p class="has-text-left is-italic">0 tracer displayed (0 total)</p>
    </div>
    <div class="column buttons is-one-half-mobile is-one-quarter-tablet is-one-quarter-desktop has-text-right">
      <button class="button is-primary"><span class="icon is-small is-left"><font-awesome-icon-stub icon="th-large"></font-awesome-icon-stub></span></button>
      <button class="button"><span class="icon is-small is-left"><font-awesome-icon-stub icon="th-list"></font-awesome-icon-stub></span></button>
    </div>
  </div>
  <tracersgrid-stub tracers="" name="fade" mode="out-in"></tracersgrid-stub>
</div>
`;
